kratos/containers/constraint_data.h

```diff
@@ -0,0 +1,407 @@
+*   This equation is imposed on the linear system of equations either element wise or on the global system.
+*
+*/
+class ConstraintEquation
+{
+  private:
+    struct MasterData
+    {
+
+        MasterData()
+        {
+            masterDofId = -1;
+            masterKey = -1;
+            masterWeight = 0.0;
+            masterEqId = -1;
+        }
+        MasterData(unsigned int iMasterDofId, unsigned int iMasterKey, double iWeight)
+        {
+            masterDofId = iMasterDofId;
+            masterKey = iMasterKey;
+            masterWeight = iWeight;
+            masterEqId = -1;
+        }
+        MasterData(unsigned int iequationId, double iWeight)
+        {
+            masterDofId = -1;
+            masterKey = -1;
+            masterEqId = iequationId;
+            masterWeight = iWeight;
+        }
+        unsigned int masterDofId;
+        unsigned int masterKey;
+        double masterWeight;
+        unsigned int masterEqId;
+    };
+
+    std::vector<MasterData> mMasterData;
+
+    unsigned int slaveDofId;
+    std::size_t slaveDofKey;
+    mutable unsigned int slaveEquationId;
+    mutable double constant;
+    mutable double constantUpdate;       
+
+  public:
+    KRATOS_CLASS_POINTER_DEFINITION(ConstraintEquation);
+    typedef std::vector<MasterData>::const_iterator const_iterator;
+    friend class ConstraintEquationContainer; 
+
+    ConstraintEquation(unsigned int iDofId, std::size_t iDofKey) : slaveDofId(iDofId), slaveDofKey(iDofKey)
+    {
+        SetConstant(0.0);
+        SetConstantUpdate(0.0);
+    }
+    ConstraintEquation(unsigned int iDofId, std::size_t iDofKey, unsigned int iEquationId) : slaveDofId(iDofId), slaveDofKey(iDofKey), slaveEquationId(iEquationId)
+    {
+        SetConstant(0.0);
+        SetConstantUpdate(0.0);
+    }
+
+/*     ConstraintEquation(const ConstraintEquation &iOther)
+    {
+        this->slaveDofId = iOther.slaveDofId;
+        this->slaveDofKey = iOther.slaveDofKey;
+        this->slaveEquationId = iOther.slaveEquationId;
+        this->constant = iOther.constant;
+        this->constantUpdate = iOther.constantUpdate;
+
+        this->mMasterData = std::move(iOther.mMasterData);
+    } */
+
+    void SetConstant(double iConstant) { constant = iConstant; }
+    void SetConstantUpdate(double iConstantUpdate) { constantUpdate = iConstantUpdate; }
+    void SetSlaveEquationId(unsigned int iEquationId) { slaveEquationId = iEquationId; }
+    double Constant() const {return constant;}
+    double ConstantUpdate() const {return constantUpdate;}
+    unsigned int SlaveDofId() const { return slaveDofId; }
+    unsigned int SlaveDofKey() const { return slaveDofKey; }
+    unsigned int SlaveEquationId() const { return slaveEquationId; }
+
+    // Add a master to this slave given are the masterDofId, masterDofKey, weight
+    void AddMasterData(unsigned int masterDofId, std::size_t masterDofKey, double weight)

```

Comment:
Here you could pass (ConstraintData, double Weight) if you allow ConstraintData to be a public class.